File: 95. Unique Binary Search Trees II/Solution.py
# LeetCode #95: Unique Binary Search Trees II

'''
Given n, generate all structurally unique BST's (binary search trees) that store values 1...n.

For example,
Given n = 3, your program should return all 5 unique BST's shown below.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
'''

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @return a list of tree node
    def generateTrees(self, n):
        return self.helper(1, n);
        
    def helper(self, start, end): # Not def helper(start, end)
        res = [];
        if start > end:
            res.append(None)
            return res
            
        for i in range(start, end+1):
            leftSubTrees = self.helper(start, i-1)
            rightSubTrees = self.helper(i+1, end)
            
            for left in leftSubTrees:
                for right in rightSubTrees:
                    root = TreeNode(i)
                    root.left = left;
                    root.right = right;
                    res.append(root);
        return res
        

File: 95. Unique Binary Search Trees II/Solution.cpp
// LeetCode #95: Unique Binary Search Trees II

/*
Given n, generate all structurally unique BST's (binary search trees) that store values 1...n.

For example,
Given n = 3, your program should return all 5 unique BST's shown below.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
*/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        return helper(1, n);
    }
    
    vector<TreeNode *> helper(int start, int end) {
        vector<TreeNode *> res;
        if (start > end) {
            res.push_back(NULL);
            return res;
        }
        
        for (int i = start; i <= end; i++) {
            vector<TreeNode *> leftSubTrees = helper(start, i-1);
            vector<TreeNode *> rightSubTrees = helper(i+1, end);
            
            for (int j = 0; j < leftSubTrees.size(); j++) {
                for (int k = 0; k < rightSubTrees.size(); k++) {
                    TreeNode* root = new TreeNode(i);
                    root->left = leftSubTrees[j];
                    root->right = rightSubTrees[k];
                    res.push_back(root);
                }
            }
        }
        return res;
    }
};